src/util/bip152.rs

```diff
@@ -0,0 +1,441 @@
+    }
+
+    /// Calculate the short ID with the given (w)txid and using the provided SipHash keys.
+    pub fn with_siphash_keys<T: AsRef<[u8]>>(txid: &T, siphash_keys: (u64, u64)) -> ShortId {
+        // 2. Running SipHash-2-4 with the input being the transaction ID and the keys (k0/k1)
+        // set to the first two little-endian 64-bit integers from the above hash, respectively.
+        let hash = siphash24::Hash::hash_with_keys(siphash_keys.0, siphash_keys.1, txid.as_ref());
+
+        // 3. Dropping the 2 most significant bytes from the SipHash output to make it 6 bytes.
+        let mut id = ShortId([0; 6]);
+        id.0.copy_from_slice(&hash[0..6]);
+        id
+    }
+}
+
+impl Encodable for ShortId {
+    #[inline]
+    fn consensus_encode<S: io::Write + ?Sized>(&self, s: &mut S) -> Result<usize, io::Error> {
+        self.0.consensus_encode(s)
+    }
+}
+
+impl Decodable for ShortId {
+    #[inline]
+    fn consensus_decode<D: io::Read + ?Sized>(d: &mut D) -> Result<ShortId, encode::Error> {
+        Ok(ShortId(Decodable::consensus_decode(d)?))
+    }
+}
+
+/// A [HeaderAndShortIds] structure is used to relay a block header, the short
+/// transactions IDs used for matching already-available transactions, and a
+/// select few transactions which we expect a peer may be missing.
+#[derive(PartialEq, Eq, Clone, Debug, PartialOrd, Ord, Hash)]
+pub struct HeaderAndShortIds {
+    /// The header of the block being provided.
+    pub header: BlockHeader,
+    ///  A nonce for use in short transaction ID calculations.
+    pub nonce: u64,
+    ///  The short transaction IDs calculated from the transactions
+    ///  which were not provided explicitly in prefilled_txs.
+    pub short_ids: Vec<ShortId>,
+    ///  Used to provide the coinbase transaction and a select few
+    ///  which we expect a peer may be missing.
+    pub prefilled_txs: Vec<PrefilledTransaction>,
+}
+impl_consensus_encoding!(HeaderAndShortIds, header, nonce, short_ids, prefilled_txs);
+
+impl HeaderAndShortIds {
+    /// Create a new [HeaderAndShortIds] from a full block.
+    ///
+    /// The version number must be either 1 or 2.
+    ///
+    /// The `prefill` slice indicates which transactions should be prefilled in
+    /// the block. It should contain the indexes in the block of the txs to
+    /// prefill. It must be ordered. 0 should not be included as the
+    /// coinbase tx is always prefilled.
+    ///
+    /// > Nodes SHOULD NOT use the same nonce across multiple different blocks.
+    pub fn from_block(
+        block: &Block,
+        nonce: u64,
+        version: u32,
+        mut prefill: &[usize],
+    ) -> Result<HeaderAndShortIds, Error> {
+        if version != 1 && version != 2 {
+            return Err(Error::UnknownVersion);
+        }
+
+        let siphash_keys = ShortId::calculate_siphash_keys(&block.header, nonce);
+
+        let mut prefilled = vec![];
+        let mut short_ids = vec![];
```

Comment:
I think we can initialize the capacity of this vectors by using `prefill.len()` and `block.txdata.len()`

Reply:
Specifically:

```rust
let mut prefilled = Vec::with_capacity(prefill.len() + 1); // +1 for coinbase tx
let mut short_ids = Vec::with_capacity(block.txdata.len() - prefill.len() - 1);
```

Reply:
I'm not sure why you subtract one in the `short_ids` capacity. This would result in a negative capacity when all transactions are pre-filled (catched by testcase 2 in `test_header_and_short_ids_from_block`). 

I think `short_ids` should be: 
```
let mut short_ids = Vec::with_capacity(block.txdata.len() - prefill.len());
```

Reply:
hmmm, so it's possible for the coinbase tx to be inside the prefill vec?

I was judging from the logic where you were `prefilled.push()` when index == 0, that coinbase tx would never be passed by the prefill vec and would always be fetched from block.txdata.

Is this behavior documented somewhere?

`prefill can contain coinbase tx or not, but either way, the coinbase tx will be treated as if it is prefilled` etc.

Reply:
I think the coinbase transaction is always pre-filled with BIP152. The receiving peer has no way of knowing the coinbase transaction before it knows the block. 

From BIP 125 in the purpose column of `prefilledtxn`:

> Used to provide the coinbase transaction and a select few which we expect a peer may be missing

https://github.com/bitcoin/bips/blob/master/bip-0152.mediawiki#headerandshortids